File: test/research_test.dart
// Research Workspace — the Dart twin of site/lib/research.test.ts.
import 'package:flutter_test/flutter_test.dart';
import 'package:crititrack/core/utils/research.dart';

final now = DateTime.utc(2026, 9, 5);
final later = DateTime.utc(2026, 9, 6);

void main() {
  group('createWorkspace', () {
    test('defaults the title from a single entity name', () {
      final w = createWorkspace(workspaceId: 'w1', userId: 'u1', entityNames: const ['Jane Doe'], now: now);
      expect(w.title, 'Research — Jane Doe');
      expect(w.status, WorkspaceStatus.active);
    });

    test('defaults to a compare title for exactly two entities', () {
      final w = createWorkspace(workspaceId: 'w1', userId: 'u1', entityNames: const ['A', 'B'], now: now);
      expect(w.title, 'Compare A and B');
    });

    test('an explicit title wins over the default', () {
      final w = createWorkspace(workspaceId: 'w1', userId: 'u1', entityNames: const ['A'], title: 'My research', now: now);
      expect(w.title, 'My research');
    });
  });

  group('workspace mutators', () {
    final base = createWorkspace(workspaceId: 'w1', userId: 'u1', entityNames: const ['Jane Doe'], now: now);

    test('rename ignores a blank title', () {
      expect(renameWorkspace(base, '   ', later).title, base.title);
      expect(renameWorkspace(base, 'New title', later).title, 'New title');
    });

    test('archive then reactivate round-trips status', () {
      final archived = archiveWorkspace(base, later);
      expect(archived.status, WorkspaceStatus.archived);
      expect(reactivateWorkspace(archived, later).status, WorkspaceStatus.active);
    });

    test('description is trimmed and bumps updatedAt', () {
      final updated = updateWorkspaceDescription(base, '  notes  ', later);
      expect(updated.description, 'notes');
      expect(updated.updatedAt, later);
    });

    test('adding the same entity twice is a no-op the second time', () {
      final once = addEntityToWorkspace(base, 'q2', later);
      final twice = addEntityToWorkspace(once, 'q2', later);
      expect(twice.entityIds, once.entityIds);
    });

    test('removing an entity not present is a no-op', () {
      expect(removeEntityFromWorkspace(base, 'q999', later), same(base));
    });
  });

  group('stableItemKey / addResearchItem deduplication', () {
    test('a NOTE (no referenceId) never collides', () {
      expect(stableItemKey(ResearchItemType.note, null), isNull);
    });

    test('adding the same (type, referenceId) twice updates in place rather than duplicating', () {
      final first = buildResearchItem(
        itemId: 'i1',
        workspaceId: 'w1',
        type: ResearchItemType.evidence,
        referenceId: 'E123',
        title: 'Old title',
        metadata: const {'confidence': 'medium'},
        now: now,
      );
      final afterFirst = addResearchItem(const [], first).items;
      expect(afterFirst, hasLength(1));

      final second = buildResearchItem(
        itemId: 'i2',
        workspaceId: 'w1',
        type: ResearchItemType.evidence,
        referenceId: 'E123',
        title: 'Refreshed title',
        metadata: const {'confidence': 'high'},
        now: later,
      );
      final result = addResearchItem(afterFirst, second);
      expect(result.added, isFalse);
      expect(result.items, hasLength(1));
      expect(result.items.first.itemId, 'i1');
      expect(result.items.first.title, 'Refreshed title');
      expect(result.items.first.metadata['confidence'], 'high');
    });

    test("a user's note and tags survive a metadata refresh", () {
      final first = buildResearchItem(
        itemId: 'i1',
        workspaceId: 'w1',
        type: ResearchItemType.claim,
        referenceId: 'C1',
        title: 'Claim A',
        note: 'keep me',
        tags: const ['important'],
        now: now,
      );
      final items = addResearchItem(const [], first).items;
      final refreshed = buildResearchItem(
        itemId: 'i2',
        workspaceId: 'w1',
        type: ResearchItemType.claim,
        referenceId: 'C1',
        title: 'Claim A (updated evidence count)',
        now: later,
      );
      final after = addResearchItem(items, refreshed).items;
      expect(after.first.note, 'keep me');
      expect(after.first.tags, ['important']);
    });

    test('different references never collapse into one item', () {
      final a = buildResearchItem(itemId: 'a', workspaceId: 'w1', type: ResearchItemType.evidence, referenceId: 'E1', title: 'A', now: now);
      final b = buildResearchItem(itemId: 'b', workspaceId: 'w1', type: ResearchItemType.evidence, referenceId: 'E2', title: 'B', now: now);
      final step1 = addResearchItem(const [], a).items;
      final step2 = addResearchItem(step1, b).items;
      expect(step2, hasLength(2));
    });

    test('two freestanding notes never dedupe against each other', () {
      final n1 = createNoteItem(itemId: 'n1', workspaceId: 'w1', text: 'first', now: now);
      final n2 = createNoteItem(itemId: 'n2', workspaceId: 'w1', text: 'second', now: now);
      final step1 = addResearchItem(const [], n1).items;
      final result = addResearchItem(step1, n2);
      expect(result.added, isTrue);
      expect(result.items, hasLength(2));
    });
  });

  group('item mutators', () {
    final item = buildResearchItem(itemId: 'i1', workspaceId: 'w1', type: ResearchItemType.claim, referenceId: 'C1', title: 'Claim A', now: now);

    test('removeResearchItem drops only the named item', () {
      final list = [item];
      expect(removeResearchItem(list, 'i1'), isEmpty);
      expect(removeResearchItem(list, 'missing'), list);
    });

    test('setItemStatus never touches other items', () {
      final other = buildResearchItem(itemId: 'i2', workspaceId: 'w1', type: ResearchItemType.note, title: 'Research note', now: now);
      final list = setItemStatus([item, other], 'i1', FindingStatus.needsReview, later);
      expect(list[0].status, FindingStatus.needsReview);
      expect(list[1].status, FindingStatus.undecided);
    });

    test('setItemNote updates the note text and updatedAt', () {
      final list = setItemNote([item], 'i1', 'Need to verify the date.', later);
      expect(list[0].note, 'Need to verify the date.');
      expect(list[0].updatedAt, later);
    });

    test('addResearchTag / removeResearchTag normalize case and de-duplicate', () {
      var list = addResearchTag([item], 'i1', 'Important', later);
      list = addResearchTag(list, 'i1', 'important', later);
      expect(list[0].tags, ['Important']);
      list = removeResearchTag(list, 'i1', 'IMPORTANT', later);
      expect(list[0].tags, isEmpty);
    });
  });

  group('search / filter / sort', () {
    final items = [
      buildResearchItem(itemId: '1', workspaceId: 'w1', type: ResearchItemType.evidence, entityId: 'qA', referenceId: 'E1', title: 'Financial report', now: DateTime.utc(2026, 1, 1)),
      buildResearchItem(itemId: '2', workspaceId: 'w1', type: ResearchItemType.claim, entityId: 'qB', referenceId: 'C1', title: 'Denial statement', note: 'check this', now: DateTime.utc(2026, 2, 1)),
      buildResearchItem(itemId: '3', workspaceId: 'w1', type: ResearchItemType.note, title: 'Research note', note: 'unrelated', tags: const ['follow-up'], now: DateTime.utc(2026, 3, 1)),
    ];

    test('search matches title, note, and tags case-insensitively', () {
      expect(searchResearchItems(items, 'financial'), hasLength(1));
      expect(searchResearchItems(items, 'check this'), hasLength(1));
      expect(searchResearchItems(items, 'follow-up'), hasLength(1));
      expect(searchResearchItems(items, ''), hasLength(3));
    });

    test('filterItemsByType / filterItemsByEntity narrow the list; null returns everything', () {
      expect(filterItemsByType(items, ResearchItemType.claim), hasLength(1));
      expect(filterItemsByType(items, null), hasLength(3));
      expect(filterItemsByEntity(items, 'qA'), hasLength(1));
      expect(filterItemsByEntity(items, null), hasLength(3));
    });

    test('sortResearchItems orders newest-first or oldest-first deterministically', () {
      final newest = sortResearchItems(items, ItemSort.newest);
      expect(newest.map((i) => i.itemId).toList(), ['3', '2', '1']);
      final oldest = sortResearchItems(items, ItemSort.oldest);
      expect(oldest.map((i) => i.itemId).toList(), ['1', '2', '3']);
    });
  });

  group('overviewCounts', () {
    test('counts each type, folding every event kind into one events bucket', () {
      final items = [
        buildResearchItem(itemId: '1', workspaceId: 'w', type: ResearchItemType.entity, title: 'A', now: now),
        buildResearchItem(itemId: '2', workspaceId: 'w', type: ResearchItemType.evidence, title: 'A', now: now),
        buildResearchItem(itemId: '3', workspaceId: 'w', type: ResearchItemType.claim, title: 'A', now: now),
        buildResearchItem(itemId: '4', workspaceId: 'w', type: ResearchItemType.controversy, title: 'A', now: now),
        buildResearchItem(itemId: '5', workspaceId: 'w', type: ResearchItemType.changeEvent, title: 'A', now: now),
        buildResearchItem(itemId: '6', workspaceId: 'w', type: ResearchItemType.source, title: 'A', now: now),
        buildResearchItem(itemId: '7', workspaceId: 'w', type: ResearchItemType.note, title: 'A', now: now),
      ];
      final counts = overviewCounts(items);
      expect(counts.entities, 1);
      expect(counts.evidence, 1);
      expect(counts.claims, 1);
      expect(counts.events, 2);
      expect(counts.sources, 1);
      expect(counts.notes, 1);
    });
  });

  group('evidenceQualitySummary', () {
    test('reads confidence/corroboration that was already saved — never computes new values', () {
      final items = [
        buildResearchItem(itemId: '1', workspaceId: 'w', type: ResearchItemType.evidence, title: 'A', metadata: const {'confidence': 'high'}, now: now),
        buildResearchItem(itemId: '2', workspaceId: 'w', type: ResearchItemType.evidence, title: 'B', metadata: const {'confidence': 'medium'}, now: now),
        buildResearchItem(itemId: '3', workspaceId: 'w', type: ResearchItemType.evidence, title: 'C', metadata: const {'confidence': 'low'}, now: now),
        buildResearchItem(itemId: '4', workspaceId: 'w', type: ResearchItemType.claim, title: 'D', metadata: const {'corroborated': true}, now: now),
      ];
      final withReview = setItemStatus(items, '2', FindingStatus.needsReview, now);
      final summary = evidenceQualitySummary(withReview);
      expect(summary.evidenceCollected, 3);
      expect(summary.highConfidence, 1);
      expect(summary.mediumConfidence, 1);
      expect(summary.lowConfidence, 1);
      expect(summary.corroboratedClaims, 1);
      expect(summary.claimsNeedingReview, 1);
    });

    test('an empty workspace reports zeros, not a crash', () {
      final summary = evidenceQualitySummary(const []);
      expect(summary.evidenceCollected, 0);
      expect(summary.highConfidence, 0);
      expect(summary.corroboratedClaims, 0);
      expect(summary.claimsNeedingReview, 0);
    });
  });

  group('activity log', () {
    test('groupActivityByDay labels today and yesterday, and buckets older days by date', () {
      final nowRef = DateTime.utc(2026, 9, 5, 12);
      final entries = [
        recordActivity(activityId: 'a1', workspaceId: 'w', kind: ActivityKind.itemAdded, summary: 'Added evidence', now: DateTime.utc(2026, 9, 5, 9)),
        recordActivity(activityId: 'a2', workspaceId: 'w', kind: ActivityKind.statusChanged, summary: 'Marked needs review', now: DateTime.utc(2026, 9, 4, 9)),
        recordActivity(activityId: 'a3', workspaceId: 'w', kind: ActivityKind.workspaceCreated, summary: 'Created workspace', now: DateTime.utc(2026, 9, 1, 9)),
      ];
      final groups = groupActivityByDay(entries, now: nowRef);
      expect(groups.map((g) => g.label).toList(), ['Today', 'Yesterday', '2026-09-01']);
    });
  });
}

File: site/lib/research.test.ts
import { describe, expect, test } from "vitest";
import {
  createWorkspace,
  renameWorkspace,
  updateDescription,
  archiveWorkspace,
  reactivateWorkspace,
  addEntityToWorkspace,
  removeEntityFromWorkspace,
  buildItem,
  addItem,
  removeItem,
  setItemStatus,
  setItemNote,
  addTag,
  removeTag,
  createNoteItem,
  searchItems,
  filterByType,
  filterByEntity,
  sortItems,
  overviewCounts,
  evidenceQualitySummary,
  recordActivity,
  groupActivityByDay,
  stableItemKey,
  type ResearchItem,
} from "./research";

const NOW = "2026-09-05T00:00:00.000Z";
const LATER = "2026-09-06T00:00:00.000Z";

describe("createWorkspace", () => {
  test("defaults the title from a single entity name", () => {
    const w = createWorkspace({ workspaceId: "w1", userId: "u1", entityNames: ["Jane Doe"], now: NOW });
    expect(w.title).toBe("Research — Jane Doe");
    expect(w.status).toBe("ACTIVE");
  });

  test("defaults to a compare title for exactly two entities", () => {
    const w = createWorkspace({ workspaceId: "w1", userId: "u1", entityNames: ["A", "B"], now: NOW });
    expect(w.title).toBe("Compare A and B");
  });

  test("an explicit title wins over the default", () => {
    const w = createWorkspace({ workspaceId: "w1", userId: "u1", entityNames: ["A"], title: "My research", now: NOW });
    expect(w.title).toBe("My research");
  });
});

describe("workspace mutators", () => {
  const base = createWorkspace({ workspaceId: "w1", userId: "u1", entityNames: ["Jane Doe"], now: NOW });

  test("rename ignores a blank title", () => {
    expect(renameWorkspace(base, "   ", LATER).title).toBe(base.title);
    expect(renameWorkspace(base, "New title", LATER).title).toBe("New title");
  });

  test("archive then reactivate round-trips status", () => {
    const archived = archiveWorkspace(base, LATER);
    expect(archived.status).toBe("ARCHIVED");
    expect(reactivateWorkspace(archived, LATER).status).toBe("ACTIVE");
  });

  test("description is trimmed and bumps updatedAt", () => {
    const updated = updateDescription(base, "  notes  ", LATER);
    expect(updated.description).toBe("notes");
    expect(updated.updatedAt).toBe(LATER);
  });

  test("adding the same entity twice is a no-op the second time", () => {
    const once = addEntityToWorkspace(base, "q2", LATER);
    const twice = addEntityToWorkspace(once, "q2", LATER);
    expect(twice.entityIds).toEqual(once.entityIds);
  });

  test("removing an entity not present is a no-op", () => {
    expect(removeEntityFromWorkspace(base, "q999", LATER)).toBe(base);
  });
});

describe("stableItemKey / addItem deduplication", () => {
  test("a NOTE (no referenceId) never collides", () => {
    expect(stableItemKey("NOTE", null)).toBeNull();
  });

  test("adding the same (type, referenceId) twice updates in place rather than duplicating", () => {
    const first = buildItem({
      itemId: "i1",
      workspaceId: "w1",
      type: "EVIDENCE",
      referenceId: "E123",
      title: "Old title",
      metadata: { confidence: "medium" },
      now: NOW,
    });
    const { items: afterFirst } = addItem([], first);
    expect(afterFirst).toHaveLength(1);

    const second = buildItem({
      itemId: "i2",
      workspaceId: "w1",
      type: "EVIDENCE",
      referenceId: "E123",
      title: "Refreshed title",
      metadata: { confidence: "high" },
      now: LATER,
    });
    const { items: afterSecond, added } = addItem(afterFirst, second);
    expect(added).toBe(false);
    expect(afterSecond).toHaveLength(1);
    expect(afterSecond[0].itemId).toBe("i1"); // identity of the retained row is the original
    expect(afterSecond[0].title).toBe("Refreshed title");
    expect(afterSecond[0].metadata.confidence).toBe("high");
  });

  test("a user's note and tags survive a metadata refresh", () => {
    const first = buildItem({
      itemId: "i1",
      workspaceId: "w1",
      type: "CLAIM",
      referenceId: "C1",
      title: "Claim A",
      note: "keep me",
      tags: ["important"],
      now: NOW,
    });
    const { items } = addItem([], first);
    const refreshed = buildItem({
      itemId: "i2",
      workspaceId: "w1",
      type: "CLAIM",
      referenceId: "C1",
      title: "Claim A (updated evidence count)",
      now: LATER,
    });
    const { items: after } = addItem(items, refreshed);
    expect(after[0].note).toBe("keep me");
    expect(after[0].tags).toEqual(["important"]);
  });

  test("different references never collapse into one item", () => {
    const a = buildItem({ itemId: "a", workspaceId: "w1", type: "EVIDENCE", referenceId: "E1", title: "A", now: NOW });
    const b = buildItem({ itemId: "b", workspaceId: "w1", type: "EVIDENCE", referenceId: "E2", title: "B", now: NOW });
    const { items: step1 } = addItem([], a);
    const { items: step2 } = addItem(step1, b);
    expect(step2).toHaveLength(2);
  });

  test("two freestanding notes never dedupe against each other", () => {
    const n1 = createNoteItem({ itemId: "n1", workspaceId: "w1", text: "first", now: NOW });
    const n2 = createNoteItem({ itemId: "n2", workspaceId: "w1", text: "second", now: NOW });
    const { items: step1 } = addItem([], n1);
    const { items: step2, added } = addItem(step1, n2);
    expect(added).toBe(true);
    expect(step2).toHaveLength(2);
  });
});

describe("item mutators", () => {
  const item = buildItem({ itemId: "i1", workspaceId: "w1", type: "CLAIM", referenceId: "C1", title: "Claim A", now: NOW });

  test("removeItem drops only the named item", () => {
    const list = [item];
    expect(removeItem(list, "i1")).toEqual([]);
    expect(removeItem(list, "missing")).toEqual(list);
  });

  test("setItemStatus never touches other items", () => {
    const other = buildItem({ itemId: "i2", workspaceId: "w1", type: "NOTE", title: "Research note", now: NOW });
    const list = setItemStatus([item, other], "i1", "NEEDS_REVIEW", LATER);
    expect(list[0].status).toBe("NEEDS_REVIEW");
    expect(list[1].status).toBe("UNDECIDED");
  });

  test("setItemNote updates the note text and updatedAt", () => {
    const list = setItemNote([item], "i1", "Need to verify the date.", LATER);
    expect(list[0].note).toBe("Need to verify the date.");
    expect(list[0].updatedAt).toBe(LATER);
  });

  test("addTag / removeTag normalize case and de-duplicate", () => {
    let list = addTag([item], "i1", "Important", LATER);
    list = addTag(list, "i1", "important", LATER); // same tag, different case
    expect(list[0].tags).toEqual(["Important"]);
    list = removeTag(list, "i1", "IMPORTANT", LATER);
    expect(list[0].tags).toEqual([]);
  });
});

describe("search / filter / sort", () => {
  const items: ResearchItem[] = [
    buildItem({ itemId: "1", workspaceId: "w1", type: "EVIDENCE", entityId: "qA", referenceId: "E1", title: "Financial report", now: "2026-01-01T00:00:00Z" }),
    buildItem({ itemId: "2", workspaceId: "w1", type: "CLAIM", entityId: "qB", referenceId: "C1", title: "Denial statement", note: "check this", now: "2026-02-01T00:00:00Z" }),
    buildItem({ itemId: "3", workspaceId: "w1", type: "NOTE", title: "Research note", note: "unrelated", tags: ["follow-up"], now: "2026-03-01T00:00:00Z" }),
  ];

  test("search matches title, note, and tags case-insensitively", () => {
    expect(searchItems(items, "financial")).toHaveLength(1);
    expect(searchItems(items, "check this")).toHaveLength(1);
    expect(searchItems(items, "follow-up")).toHaveLength(1);
    expect(searchItems(items, "")).toHaveLength(3);
  });

  test("filterByType / filterByEntity narrow the list; ALL returns everything", () => {
    expect(filterByType(items, "CLAIM")).toHaveLength(1);
    expect(filterByType(items, "ALL")).toHaveLength(3);
    expect(filterByEntity(items, "qA")).toHaveLength(1);
    expect(filterByEntity(items, "ALL")).toHaveLength(3);
  });

  test("sortItems orders newest-first or oldest-first deterministically", () => {
    const newest = sortItems(items, "newest");
    expect(newest.map((i) => i.itemId)).toEqual(["3", "2", "1"]);
    const oldest = sortItems(items, "oldest");
    expect(oldest.map((i) => i.itemId)).toEqual(["1", "2", "3"]);
  });
});

describe("overviewCounts", () => {
  test("counts each type, folding every event kind into one events bucket", () => {
    const items: ResearchItem[] = [
      buildItem({ itemId: "1", workspaceId: "w", type: "ENTITY", title: "A", now: NOW }),
      buildItem({ itemId: "2", workspaceId: "w", type: "EVIDENCE", title: "A", now: NOW }),
      buildItem({ itemId: "3", workspaceId: "w", type: "CLAIM", title: "A", now: NOW }),
      buildItem({ itemId: "4", workspaceId: "w", type: "CONTROVERSY", title: "A", now: NOW }),
      buildItem({ itemId: "5", workspaceId: "w", type: "CHANGE_EVENT", title: "A", now: NOW }),
      buildItem({ itemId: "6", workspaceId: "w", type: "SOURCE", title: "A", now: NOW }),
      buildItem({ itemId: "7", workspaceId: "w", type: "NOTE", title: "A", now: NOW }),
    ];
    expect(overviewCounts(items)).toEqual({
      entities: 1,
      evidence: 1,
      claims: 1,
      events: 2,
      sources: 1,
      notes: 1,
    });
  });
});

describe("evidenceQualitySummary", () => {
  test("reads confidence/corroboration that was already saved — never computes new values", () => {
    const items: ResearchItem[] = [
      buildItem({ itemId: "1", workspaceId: "w", type: "EVIDENCE", title: "A", metadata: { confidence: "high" }, now: NOW }),
      buildItem({ itemId: "2", workspaceId: "w", type: "EVIDENCE", title: "B", metadata: { confidence: "medium" }, now: NOW }),
      buildItem({ itemId: "3", workspaceId: "w", type: "EVIDENCE", title: "C", metadata: { confidence: "low" }, now: NOW }),
      buildItem({ itemId: "4", workspaceId: "w", type: "CLAIM", title: "D", metadata: { corroborated: true }, now: NOW }),
    ];
    const withReview = setItemStatus(items, "2", "NEEDS_REVIEW", NOW);
    const summary = evidenceQualitySummary(withReview);
    expect(summary.evidenceCollected).toBe(3);
    expect(summary.highConfidence).toBe(1);
    expect(summary.mediumConfidence).toBe(1);
    expect(summary.lowConfidence).toBe(1);
    expect(summary.corroboratedClaims).toBe(1);
    expect(summary.claimsNeedingReview).toBe(1);
  });

  test("an empty workspace reports zeros, not a crash", () => {
    expect(evidenceQualitySummary([])).toEqual({
      evidenceCollected: 0,
      highConfidence: 0,
      mediumConfidence: 0,
      lowConfidence: 0,
      corroboratedClaims: 0,
      claimsNeedingReview: 0,
    });
  });
});

describe("activity log", () => {
  test("groupActivityByDay labels today and yesterday, and buckets older days by date", () => {
    const now = new Date("2026-09-05T12:00:00.000Z");
    const entries = [
      recordActivity({ activityId: "a1", workspaceId: "w", kind: "item_added", summary: "Added evidence", now: "2026-09-05T09:00:00.000Z" }),
      recordActivity({ activityId: "a2", workspaceId: "w", kind: "status_changed", summary: "Marked needs review", now: "2026-09-04T09:00:00.000Z" }),
      recordActivity({ activityId: "a3", workspaceId: "w", kind: "workspace_created", summary: "Created workspace", now: "2026-09-01T09:00:00.000Z" }),
    ];
    const groups = groupActivityByDay(entries, now);
    expect(groups.map((g) => g.label)).toEqual(["Today", "Yesterday", "2026-09-01"]);
  });
});
